ste/xfer-deleteFile.go

```diff
@@ -11,24 +12,77 @@ import (
+	// as much of the plan files as we have seen so far)
+	// That minimizes case where the count of known children drops to zero (due simply to us
+	// having not registered all of them yet); and the manager attempts a failed deletion;
+	// and then we find more children in the plan files. Such failed attempts are harmless, but cause
+	// unnecessary network round trips.
+	// We must do this for all entity types, because even folders are children of their parents
+	jptm.FolderDeletionManager().RecordChildExists(srcUrl)
+
+	if info.EntityType == common.EEntityType.Folder() {
+
+		au := azfile.NewFileURLParts(*srcUrl)
+		isFileShareRoot := au.DirectoryOrFilePath == ""
+		if !isFileShareRoot {
+			jptm.LogAtLevelForCurrentTransfer(pipeline.LogInfo, "Queuing folder, to be deleted after it's children are deleted")
+			jptm.FolderDeletionManager().RequestDeletion(
+				srcUrl,
+				func(ctx context.Context, logger common.ILogger) bool {
+					return doDeleteFolder(ctx, info.Source, p, logger)
+				},
+			)
+		}
+		// After requesting deletion, we have no choice but to report this as "done", because we are
+		// in a transfer initiation func, and can't just block here for ages until the deletion actually happens.
+		// Besides, we have made the decision that if the queued deletion fails, that's NOT a
+		// job failure. (E.g. could happen because someone else dropped a new file
+		// in there after we enumerated). Since the deferred action (by this definition)
+		// will never fail, it's correct to report success here.
+		jptm.SetStatus(common.ETransferStatus.Success())
+		jptm.ReportTransferDone()
+
+	} else {
+		// schedule the work as a chunk, so it will run on the main goroutine pool, instead of the
+		// smaller "transfer initiation pool", where this code runs.
+		id := common.NewChunkID(info.Source, 0, 0)
+		cf := createChunkFunc(true, jptm, id, func() { doDeleteFile(jptm, p) })
+		jptm.ScheduleChunks(cf)
+	}
+}
+
+func doDeleteFile(jptm IJobPartTransferMgr, p pipeline.Pipeline) {
+
+	info := jptm.Info()
+	// Get the source file url of file to delete
+	srcUrl, _ := url.Parse(info.Source)
+
+	srcFileUrl := azfile.NewFileURL(*srcUrl, p)
+
 	// Internal function which checks the transfer status and logs the msg respectively.
 	// Sets the transfer status and Report Transfer as Done.
 	// Internal function is created to avoid redundancy of the above steps from several places in the api.
 	transferDone := func(status common.TransferStatus, err error) {
+		if status == common.ETransferStatus.Success() {
+			jptm.FolderDeletionManager().RecordChildDeleted(srcUrl)
+			// TODO: doing this only on success raises the possibility of the
+			//   FolderDeletionManager's internal may growing rather large if there are lots of failures
+			//   on a big folder tree. Is living with that preferable to the "incorrectness" of calling
+			//   RecordChildDeleted when it wasn't actually deleted.  Yes, probably.  But think about it a bit more.
+		}
```

Comment:
I'm somewhere in between.

We don't record the folder deletion as a transfer failure, so it wouldn't hurt to call this and have the folder deletion fail

Alternative naming suggestion:

RecordChildDeletionAttempted

Reply:
Let's favor correctness over memory-efficiency for now, and leave the code as it is.  If we find that memory usage is an issue in cases with lots of failures, we can revisit.

I have updated the TODO a little.